src/Compilers/Core/Portable/SourceGeneration/Nodes/GeneratorRunStateTable.cs

```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class GeneratorRunStateTable
+    {
+        private GeneratorRunStateTable(ImmutableDictionary<string, ImmutableArray<IncrementalGeneratorRunStep>> executedSteps, ImmutableDictionary<string, ImmutableArray<IncrementalGeneratorRunStep>> outputSteps)
+        {
+            ExecutedSteps = executedSteps;
+            OutputSteps = outputSteps;
+        }
+
+        public ImmutableDictionary<string, ImmutableArray<IncrementalGeneratorRunStep>> ExecutedSteps { get; }
+
+        public ImmutableDictionary<string, ImmutableArray<IncrementalGeneratorRunStep>> OutputSteps { get; }
+
+        public class Builder
+        {
+            private readonly Dictionary<string, HashSet<IncrementalGeneratorRunStep>>? _namedSteps;
+            private readonly Dictionary<string, HashSet<IncrementalGeneratorRunStep>>? _outputSteps;
+
+            public Builder(bool recordingExecutedSteps)
+            {
+                RecordingExecutedSteps = recordingExecutedSteps;
+                if (RecordingExecutedSteps)
+                {
+                    _namedSteps = new Dictionary<string, HashSet<IncrementalGeneratorRunStep>>();
+                    _outputSteps = new Dictionary<string, HashSet<IncrementalGeneratorRunStep>>();
+                }
+            }
+
+            public void RecordStepsFromOutputNodeUpdate(IStateTable table)
+            {
+                Debug.Assert(RecordingExecutedSteps);
+                Debug.Assert(table.HasTrackedSteps);
+                foreach (var step in table.Steps)
+                {
+                    RecordStepTree(step, addToOutputSteps: true);
+                }
+            }
+
+            [MemberNotNull(nameof(_namedSteps), nameof(_outputSteps))]
+            public bool RecordingExecutedSteps { get; }
+
+            public GeneratorRunStateTable ToImmutableAndFree()
+            {
+                if (!RecordingExecutedSteps)
```

Comment:
>RecordingExecutedSteps [](http://example.com/codeflow?start=21&length=22)

Could `StepCollectionToImmutableAndFree()` simply return an empty dictionary if the argument is `null`?